src/protocol/step/compact.rs

```diff
@@ -0,0 +1,190 @@
+        return QUERY_TYPE_SEMIHONEST_STATE;
+    } else if step == crate::helpers::query::QueryType::MALICIOUS_IPA_STR {
+        return QUERY_TYPE_MALICIOUS_STATE;
+    } else if step == crate::helpers::prss_protocol::PrssExchangeStep.as_ref() {
+        return PRSS_EXCHANGE_STATE;
+    }
+
+    panic!("cannot narrow with \"{step}\" from state {state}");
+}
+
+// Reverse of `static_state_map` for `Compact::as_ref()`
+fn static_reverse_state_map(state: u16) -> &'static str {
+    match state {
+        ROOT_STATE => "run-0",
+        FALLBACK_STATE => crate::protocol::boolean::random_bits_generator::FallbackStep.as_ref(),
+        UPGRADE_SEMI_HONEST_STATE => crate::protocol::context::semi_honest::UpgradeStep.as_ref(),
+        QUERY_TYPE_SEMIHONEST_STATE => crate::helpers::query::QueryType::SEMIHONEST_IPA_STR,
+        QUERY_TYPE_MALICIOUS_STATE => crate::helpers::query::QueryType::MALICIOUS_IPA_STR,
+        PRSS_EXCHANGE_STATE => crate::helpers::prss_protocol::PrssExchangeStep.as_ref(),
+        _ => panic!("cannot as_ref() from the invalid state {state}"),
+    }
+}
+
+fn static_deserialize_state_map(s: &str) -> u16 {
+    if s == crate::protocol::boolean::random_bits_generator::FallbackStep.as_ref() {
+        return FALLBACK_STATE;
+    } else if s == crate::protocol::context::semi_honest::UpgradeStep.as_ref() {
+        return UPGRADE_SEMI_HONEST_STATE;
+    } else if s == crate::helpers::query::QueryType::SEMIHONEST_IPA_STR {
+        return QUERY_TYPE_SEMIHONEST_STATE;
+    } else if s == crate::helpers::query::QueryType::MALICIOUS_IPA_STR {
+        return QUERY_TYPE_MALICIOUS_STATE;
+    } else if s == crate::helpers::prss_protocol::PrssExchangeStep.as_ref() {
+        return PRSS_EXCHANGE_STATE;
+    }
+
+    panic!("cannot deserialize from the invalid step \"{s}\"");
+}
+
+//
+// "conditional" steps
+//
+
+impl StepNarrow<crate::protocol::boolean::random_bits_generator::FallbackStep> for Compact {
+    // `src/protocol/boolean/random_bits_generator.rs`
+    // We create `fallback_ctx` because the underlying bits generator can fail. In reality,
+    // this only happens in 5/2^32 cases with `Fp32BitPrime` so we need to rely on the static map.
+    fn narrow(&self, step: &crate::protocol::boolean::random_bits_generator::FallbackStep) -> Self {
+        Self(static_state_map(self.0, step.as_ref()))
+    }
+}
+
+// steps that do not trigger communications
+
+impl StepNarrow<crate::protocol::context::semi_honest::UpgradeStep> for Compact {
+    // `src/protocol/context/semi_honest.rs`
+    // Semi-honest implementations of `UpgradedContext::upgrade()` and subsequent
+    // `UpgradeToMalicious::upgrade()` narrows but these will end up in `UpgradedContext::upgrade_one()`
+    // or `UpgradedContext::upgrade_sparse()` which both return Ok() and never trigger communications.
+    fn narrow(&self, step: &crate::protocol::context::semi_honest::UpgradeStep) -> Self {
+        Self(static_state_map(self.0, step.as_ref()))
```

Comment:
I wanted to hide these inside compact.rs because, as you said, these are somewhat "magic". These are kind of temporary solution until we have a new macro that supports annotations for "not-executed steps". Defining a blanket implementation doesn't sound right to me for the same reason because descriptive gate also implements StepNarrow trait.

Reply:
oh what I meant was something along these lines

```rust
trait NoCommsStep : AsRef<str> {}

impl <C: Step + NoCommsStep> StepNarrow<C> for Compact {
    fn narrow(&self, step: &C) -> Self {
        Self(static_state_map(self.0, step.as_ref()))
    }
}

impl NoCommsStep for crate::protocol::context::semi_honest::UpgradeStep {}
impl NoCommsStep for crate::protocol::query::QueryType {}
```

just cosmetics but helps keeping `static_state_map` usage in one place

Reply:
Ah I see. Thanks! Modified as suggested.